tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,553 @@
+package tensorflow
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+
+	pb "github.com/tensorflow/tensorflow/tensorflow/contrib/go/proto"
+)
+
+const (
+	cOpsProtobufDefsPath = "/usr/local/tensorlow/ops.pbtxt"
+)
+
+// A Graph is the representation of the computation graph.
+type Graph struct {
+	def *pb.GraphDef
+
+	availableOps map[string]*pb.OpDef
+	constants    map[string]*Tensor
+	variables    map[string]*Tensor
+}
+
+// A GraphNode is the representation of one of the nodes of the TensorFlow
+// Graph. A node takes zero or more Tensors, performs some computation, and
+// produces zero or more Tensors.
+type GraphNode struct {
+	ref          *pb.NodeDef
+	def          *pb.NodeDef
+	outDataTypes map[string]DataType
+}
+
+// NewGraph returns an initialized instance of the Graph struct.
+func NewGraph() *Graph {
+	return &Graph{
+		def:          new(pb.GraphDef),
+		availableOps: make(map[string]*pb.OpDef),
+		constants:    make(map[string]*Tensor),
+		variables:    make(map[string]*Tensor),
+	}
+}
+
+// NewGraphFromReader reads from reader until an error or EOF and loads the
+// content into a new graph. Use the asText parameter to specify if the graph
+// from the reader is provided in Text format.
+func NewGraphFromReader(reader io.Reader, asText bool) (gr *Graph, err error) {
```

Comment:
given the api you'd need to allocate with string([]byte()) -- I think we should err on not allocating and accepting a string directly.